src/Language/PureScript/Sugar/Names.hs

```diff
@@ -137,17 +139,22 @@ elaborateImports imps (Module ss coms mn decls exps) = Module ss coms mn decls'
   mkImport :: ModuleName -> Declaration
   mkImport mn' = ImportDeclaration mn' (Explicit []) Nothing
 
+
+
 -- |
 -- Replaces all local names with qualified names within a module and checks that all existing
 -- qualified names are valid.
 --
-renameInModule :: forall m. (Applicative m, MonadError MultipleErrors m) => Env -> Imports -> Module -> m Module
-renameInModule env imports (Module ss coms mn decls exps) =
-  Module ss coms mn <$> parU decls go <*> pure exps
+renameInModule :: forall m. (Applicative m, MonadError MultipleErrors m, MonadWriter MultipleErrors m) => Env -> Imports -> Module -> m Module
+renameInModule env imports m@(Module ss coms mn decls exps) = do
+  (decls', used) <- flip runStateT M.empty $ parU decls go
+  findUnusedImports m imports used
+  return $ Module ss coms mn decls' exps
   where
+
   (go, _, _, _, _) = everywhereWithContextOnValuesM (Nothing, []) updateDecl updateValue updateBinder updateCase defS
 
-  updateDecl :: (Maybe SourceSpan, [Ident]) -> Declaration -> m ((Maybe SourceSpan, [Ident]), Declaration)
+  updateDecl :: (Maybe SourceSpan, [Ident]) -> Declaration -> StateT UsedImports m ((Maybe SourceSpan, [Ident]), Declaration)
```

Comment:
Instead of using `StateT` explicitly everywhere here, why not add a `MonadState UsedImports m` constraint to `renameInModule`? You won't have to `lift` the state operations then either.